src/puppetlabs/puppetdb/command.clj

```diff
@@ -778,110 +796,156 @@
-             :consumer-threadpool command-pool
-             :broadcast-threadpool broadcast-pool))))
+
+      ;; FIXME: check that nil command-chan is actually fatal if appropriate
+      (let [shovel #(with-monitored-execution shutdown-for-ex
+                      (try
+                        (pool/dochan command-pool handle-cmd command-chan)
+                        (catch ExceptionInfo ex
+                          (when-not (= :puppetlabs.puppetdb.threadpool/rejected
+                                       (:kind (ex-data ex)))
+                            (throw ex)))
+                        (catch InterruptedException ex
+                          nil)))
+            shovel-thread (doto (Thread. shovel)
+                            (.setName "PuppetDB command shovel")
+                            (.setDaemon false)
+                            (.start))]
+        (assoc context
+               :command-shovel shovel-thread
+               :command-chan command-chan
+               :consumer-threadpool command-pool
+               :broadcast-threadpool broadcast-pool)))))
 
 (defn stop-command-service
   [{:keys [stop-status consumer-threadpool broadcast-threadpool command-chan
-           delay-pool]
+           response-chan response-chan-for-pub response-mult response-pub
+           delay-pool command-shovel]
     :as context}]
-  (some-> command-chan async/close!)
-  ;; FIXME: (PDB-4742) exception suppression?
-  (some-> consumer-threadpool pool/shutdown-gated)
-  (some-> broadcast-threadpool pool/shutdown-unbounded)
-  (some-> delay-pool stop-and-reset-pool!)
-  ;; Wait up to ~5s for https://bugs.openjdk.java.net/browse/JDK-8176254
-  (swap! stop-status #(assoc % :stopping true))
-  (if (utils/wait-for-ref-state stop-status (stop-commands-wait-ms)
-                                #(= % {:stopping true :executing-delayed 0}))
-    (log/info (trs "Halted delayed command processsing"))
-    (log/info (trs "Forcibly terminating delayed command processing")))
-  (async/unsub-all (:response-pub context))
-  (async/untap-all (:response-mult context))
-  (async/close! (:response-chan-for-pub context))
-  (async/close! (:response-chan context))
-  (dissoc context
-          :response-pub :response-chan :response-chan-for-pub :response-mult))
+  (try!
+   (dissoc context
+           :response-pub :response-chan :response-chan-for-pub
+           :response-mult)
+   ;; The reason some of these aren't conditional is because they're
+   ;; established by init, tk won't call stop if init didn't
+   ;; finish, and in those cases a nil would indicate a bug.
+   (finally (some-> command-chan async/close!))
+   (finally (when command-shovel (.interrupt command-shovel)))
+   (finally (some-> broadcast-threadpool pool/shutdown-unbounded))
+   (finally (some-> consumer-threadpool pool/shutdown-gated))
+   (finally (stop-and-reset-pool! delay-pool))
+   (finally
+     ;; Wait up to ~5s for https://bugs.openjdk.java.net/browse/JDK-8176254
```

Comment:
We should ticket this for removal in Platform 7 as this bug doesn't affect Java 7